router/internal/persistedoperation/apq/client.go

```diff
@@ -0,0 +1,94 @@
+package apq
+
+import (
+	"context"
+	"errors"
+	"github.com/wundergraph/cosmo/router/internal/persistedoperation/operationstorage"
+	"github.com/wundergraph/cosmo/router/pkg/config"
+	"go.uber.org/zap"
+	"sync"
+)
+
+type PersistedOperation struct {
+	Version int    `json:"version"`
+	Body    string `json:"body"`
+}
+
+type Client interface {
+	Enabled() bool
+	PersistedOperation(ctx context.Context, clientName string, sha256Hash string) ([]byte, error)
+	SaveOperation(ctx context.Context, clientName, sha256Hash string, operationBody []byte) error
+	Close()
+}
+
+type KVClient interface {
+	Get(ctx context.Context, clientName, operationHash string) ([]byte, error)
+	Set(ctx context.Context, operationHash string, operationBody []byte, ttl int) error
+	Close()
+}
+
+type Options struct {
+	Logger    *zap.Logger
+	ApqConfig *config.AutomaticPersistedQueriesConfig
+	KVClient  KVClient
+}
+
+type client struct {
+	enabled   bool
+	ttl       int
+	cache     *operationstorage.OperationsCache
+	kvClient  KVClient
+	cacheLock *sync.RWMutex
+}
+
+func NewClient(opts *Options) (Client, error) {
+	if opts.ApqConfig == nil {
+		return nil, errors.New("APQ config is nil")
+	}
+	cl := &client{
+		enabled:  opts.ApqConfig.Enabled,
+		kvClient: opts.KVClient,
+		ttl:      opts.ApqConfig.Cache.TTL,
+	}
+
+	if opts.ApqConfig == nil {
+		return nil, errors.New("APQ config is nil")
+	} else if !opts.ApqConfig.Enabled || opts.KVClient != nil {
+		return cl, nil
+	}
+
+	var err error
+	cl.cache, err = operationstorage.NewOperationsCache(int64(opts.ApqConfig.Cache.Size.Uint64()))
+	cl.cacheLock = &sync.RWMutex{}
+
+	return cl, err
+}
+
+func (c client) Enabled() bool {
+	return c.enabled
+}
+
+func (c client) PersistedOperation(ctx context.Context, clientName string, sha256Hash string) ([]byte, error) {
```

Comment:
I would use pointer receiver instead of value receiver (c *client) as a default. Otherwise each function call creates a client object with copies of all values, which might result in unintended behaviour, e.g. if you want to change something on the client object.